### PR TITLE
[Py] Ngraph Py API TensorIterator

### PR DESCRIPTION
**Pull request by arogowie-intel**
Additionally in this PR I've:

* Updated `DetectionOutput`, `Proposal`, `Interpolate`, `PriorBox` and `PriorBoxClustered` visit attributes method to new version of visitor API.

### CODE REVIEW

File: ngraph/python/src/ngraph/ops.py [hunk 1]
```diff
@@ -3438,3 +3446,49 @@ def proposal(
...
+    attributes = {
+        "body": graph_body.serialize(),
+        "slice_input_desc": [desc.serialize() for desc in slice_input_desc],
+        "merged_input_desc": [desc.serialize() for desc in merged_input_desc],
+        "invariant_input_desc": [desc.serialize() for desc in invariant_input_desc],
+        "body_output_desc": [desc.serialize() for desc in body_output_desc],
+        "concat_output_desc": [desc.serialize() for desc in concat_output_desc],
+    }
+
+    return _get_node_factory().create('TensorIterator', inputs, attributes)
```
Comment (postrational) [inline ngraph/python/src/ngraph/ops.py:3494]:
```suggestion
    return _get_node_factory().create('TensorIterator', as_nodes(*inputs), attributes)
```

File: ngraph/python/src/ngraph/utils/tensor_iterator_types.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+    def serialize(self) -> dict:
+        return {
+            "input_idx": self.input_idx,
+            "body_parameter_idx": self.body_parameter_idx,
+        }
+
+
+class TensorIteratorSliceInputDesc(TensorIteratorInputDesc):
+    """Represents a TI graph body input formed from slicec of an TI input."""
```
Comment (postrational) [inline ngraph/python/src/ngraph/utils/tensor_iterator_types.py:61]: Let's avoid using `TI` abbreviation.

File: ngraph/python/src/ngraph/utils/tensor_iterator_types.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        self,
+        input_idx: int,
+        body_parameter_idx: int,
+        start: int,
+        stride: int,
+        part_size: int,
+        end: int,
+        axis: int,
+    ) -> None:
+        super(TensorIteratorSliceInputDesc, self).__init__(input_idx, body_parameter_idx)
```
Comment (postrational) [inline ngraph/python/src/ngraph/utils/tensor_iterator_types.py:73]: Python 3 makes this simpler:
```suggestion
        super().__init__(input_idx, body_parameter_idx)
```

File: ngraph/python/src/ngraph/utils/tensor_iterator_types.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        self.start = start
+        self.stride = stride
+        self.part_size = part_size
+        self.end = end
+        self.axis = axis
+
+    def serialize(self) -> dict:
+        output = super(TensorIteratorSliceInputDesc, self).serialize()
```
Comment (postrational) [inline ngraph/python/src/ngraph/utils/tensor_iterator_types.py:81]:
```suggestion
        output = super().serialize()
```

File: ngraph/python/src/ngraph/utils/tensor_iterator_types.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    Later on, this input value is computed inside graph body.
+    """
+
+    def __init__(
+        self,
+        input_idx: int,
+        body_parameter_idx: int,
+        body_value_idx: int,
+    ) -> None:
+        super(TensorIteratorMergedInputDesc, self).__init__(input_idx, body_parameter_idx)
```
Comment (postrational) [inline ngraph/python/src/ngraph/utils/tensor_iterator_types.py:102]:
```suggestion
        super().__init__(input_idx, body_parameter_idx)
```

File: ngraph/python/src/ngraph/utils/tensor_iterator_types.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        self.body_value_idx = body_value_idx
+
+    def serialize(self) -> dict:
+        output = super(TensorIteratorMergedInputDesc, self).serialize()
```
Comment (postrational) [inline ngraph/python/src/ngraph/utils/tensor_iterator_types.py:106]:
```suggestion
        output = super().serialize()
```

File: ngraph/python/src/ngraph/utils/tensor_iterator_types.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+class TensorIteratorInvariantInputDesc(TensorIteratorInputDesc):
+    """Represents a TI graph body input that has invariant value during iteration."""
+
+    def __init__(
+        self,
+        input_idx: int,
+        body_parameter_idx: int,
+    ) -> None:
+        super(TensorIteratorInvariantInputDesc, self).__init__(input_idx, body_parameter_idx)
```
Comment (postrational) [inline ngraph/python/src/ngraph/utils/tensor_iterator_types.py:119]:
```suggestion
        super().__init__(input_idx, body_parameter_idx)
```

File: ngraph/python/src/ngraph/utils/tensor_iterator_types.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+
+class TensorIteratorInvariantInputDesc(TensorIteratorInputDesc):
+    """Represents a TI graph body input that has invariant value during iteration."""
+
+    def __init__(
+        self,
+        input_idx: int,
+        body_parameter_idx: int,
+    ) -> None:
```
Comment (postrational) [inline ngraph/python/src/ngraph/utils/tensor_iterator_types.py:118]: Why don't we put these signatures on one line? Try running `black --line-length=100` on this file.
```suggestion
    def __init__(self, input_idx: int, body_parameter_idx: int) -> None:
```

File: ngraph/python/src/pyngraph/tensor_iterator_builder.cpp [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+    {
+        m_merged_input_desc = m_attributes["merged_input_desc"].cast<py::list>();
+    }
+
+    if (m_attributes.contains("invariant_input_desc"))
+    {
+        m_invariant_input_desc = m_attributes["invariant_input_desc"].cast<py::list>();
+    }
+
+    // m_outputs.resize(n_outputs);
```
Comment (postrational) [inline ngraph/python/src/pyngraph/tensor_iterator_builder.cpp:47]: Leftover line.

File: ngraph/python/test/ngraph/test_create_op.py [hunk 10]
```diff
@@ -537,18 +536,6 @@ def test_convert_like():
     assert node.get_output_element_type(0) == Type.i8
 
 
-def test_one_hot():
```
Comment (postrational) [inline ngraph/python/test/ngraph/test_create_op.py:540]: Why are you removing these tests?

File: ngraph/src/ngraph/op/proposal.cpp [hunk 11]
```diff
@@ -92,19 +92,32 @@ shared_ptr<Node> op::Proposal::clone_with_new_inputs(const OutputVector& new_arg
...
+}
+
+constexpr DiscreteTypeInfo AttributeAdapter<op::ProposalAttrs>::type_info;
+
+AttributeAdapter<op::ProposalAttrs>::AttributeAdapter(op::ProposalAttrs& ref)
+    : m_ref(ref)
+{
+}
+
+bool AttributeAdapter<op::ProposalAttrs>::visit_attributes(AttributeVisitor& visitor)
```
Comment (postrational) [inline ngraph/src/ngraph/op/proposal.cpp:106]: Should this be part of this PR?

File: ngraph/python/test/ngraph/test_create_op.py (hunk 10 above)
Comment (arogowie-intel) [reply ngraph/python/test/ngraph/test_create_op.py:540]: It's duplicated here: https://github.com/arogowie-intel/openvino/blob/aosewski/ng_py_tensor_iterator/ngraph/python/test/ngraph/test_sequence_processing.py#L34-L43

File: ngraph/python/src/ngraph/ops.py (hunk 1 above)
Comment (arogowie-intel) [reply ngraph/python/src/ngraph/ops.py:3494]: Is it necessary? `inputs` is has already type marked as: `List[Node]`.

File: ngraph/src/ngraph/op/proposal.cpp [hunk 12]
```diff
@@ -92,19 +92,32 @@ shared_ptr<Node> op::Proposal::clone_with_new_inputs(const OutputVector& new_arg
...
+{
+    visitor.on_attribute("base_size", m_ref.base_size);
```
[the remark this answers is not in the document]
Comment (arogowie-intel) [reply ngraph/src/ngraph/op/proposal.cpp:108]: I'll do that in separate PR.

**Pull request merged**